File: src/ts/test/idempotency-scope.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { readFile } from "node:fs/promises";
import { scopedIdempotencyKey } from "../dist/idempotency-scope.js";
const corpus = JSON.parse(await readFile(new URL("../../../contracts/fixtures/idempotency-scope.json", import.meta.url), "utf8"));
for (const item of corpus.cases) {
  test(`shared replay-scope vector: ${item.name}`, async () => {
    assert.equal(await scopedIdempotencyKey(item.scope), item.expectedKey);
  });
}
test("all intended replay scopes are distinct", async () => {
  const keys = await Promise.all(corpus.cases.map(item => scopedIdempotencyKey(item.scope)));
  assert.equal(new Set(keys).size, keys.length);
});
test("malformed replay scopes fail closed rather than silently aliasing", async () => {
  const base = corpus.cases[0].scope;
  for (const scope of [{ ...base, userId: "\ud800" }, { ...base, path: "relative" }, { ...base, idempotencyKey: "" }, { ...base, tenantId: 1 }, { ...base, extra: true }]) {
    await assert.rejects(scopedIdempotencyKey(scope), TypeError);
  }
});

File: src/golang/idempotency_scope_test.go
package oresmiddleware

import (
	"encoding/json"
	"os"
	"testing"
)

func TestSharedIdempotencyScopeVectors(t *testing.T) {
	data, err := os.ReadFile("../../contracts/fixtures/idempotency-scope.json")
	if err != nil {
		t.Fatal(err)
	}
	var corpus struct {
		Cases []struct {
			Name        string           `json:"name"`
			Scope       IdempotencyScope `json:"scope"`
			ExpectedKey string           `json:"expectedKey"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(data, &corpus); err != nil {
		t.Fatal(err)
	}
	if len(corpus.Cases) != 13 {
		t.Fatalf("expected 13 independent vectors, got %d", len(corpus.Cases))
	}
	seen := map[string]bool{}
	for _, item := range corpus.Cases {
		t.Run(item.Name, func(t *testing.T) {
			key, err := ScopedIdempotencyKey(item.Scope)
			if err != nil {
				t.Fatal(err)
			}
			if key != item.ExpectedKey {
				t.Fatalf("scope digest mismatch: %s", item.Name)
			}
			if seen[key] {
				t.Fatal("distinct scope collision")
			}
			seen[key] = true
		})
	}
}

func TestMalformedIdempotencyScopeRefused(t *testing.T) {
	base := IdempotencyScope{ServiceName: "service", TenantID: "tenant", UserID: "user", Method: "POST", Path: "/", IdempotencyKey: "key"}
	for _, scope := range []IdempotencyScope{
		{ServiceName: "service", Method: "POST", Path: "/", IdempotencyKey: ""},
		{ServiceName: "service", Method: "POST", Path: "relative", IdempotencyKey: "key"},
		{ServiceName: "service", Method: "POST", Path: "/", IdempotencyKey: string([]byte{0xff})},
	} {
		if _, err := ScopedIdempotencyKey(scope); err == nil {
			t.Fatal("malformed scope accepted")
		}
	}
	if _, err := ScopedIdempotencyKey(base); err != nil {
		t.Fatal(err)
	}
}
